File: lib/ftw_realtime_web/controllers/api/job_controller.ex
defmodule FtwRealtimeWeb.Api.JobController do
  use FtwRealtimeWeb, :controller

  alias FtwRealtime.Marketplace

  def index(conn, params) do
    limit = case Integer.parse(params["limit"] || "") do
      {n, _} when n > 0 -> n
      _ -> 20
    end

    opts = [
      status: params["status"],
      category: params["category"],
      limit: limit,
      after: params["after"]
    ]

    %{jobs: jobs, has_more: has_more, next_cursor: next_cursor} = Marketplace.list_jobs(opts)

    json(conn, %{
      jobs: Enum.map(jobs, &serialize_job/1),
      has_more: has_more,
      next_cursor: next_cursor
    })
  end

  def show(conn, %{"id" => id}) do
    case Marketplace.get_job(id) do
      nil ->
        conn |> put_status(:not_found) |> json(%{error: "Job not found"})

      job ->
        bids = Marketplace.list_bids(id)

        json(conn, %{
          job: serialize_job(job),
          bids: Enum.map(bids, &serialize_bid/1)
        })
    end
  end

  def create(conn, %{"job" => job_params}) do
    job_params = Map.put(job_params, "homeowner_id", conn.assigns.current_user_id)

    case Marketplace.post_job(job_params) do
      {:ok, job} ->
        conn |> put_status(:created) |> json(%{job: serialize_job(job)})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  def place_bid(conn, %{"id" => job_id, "bid" => bid_params}) do
    bid_params = Map.put(bid_params, "contractor_id", conn.assigns.current_user_id)

    case Marketplace.place_bid(job_id, bid_params) do
      {:ok, bid} ->
        conn |> put_status(:created) |> json(%{bid: serialize_bid(bid)})

      {:error, :job_not_found} ->
        conn |> put_status(:not_found) |> json(%{error: "Job not found"})

      {:error, :job_not_accepting_bids} ->
        conn |> put_status(:conflict) |> json(%{error: "Job is not accepting bids"})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  def transition(conn, %{"id" => job_id, "status" => status_str}) do
    status = String.to_existing_atom(status_str)

    case Marketplace.transition_job(job_id, status, conn.assigns.current_user_id) do
      {:ok, job} ->
        json(conn, %{job: serialize_job(job)})

      {:error, {:invalid_transition, from, to}} ->
        conn
        |> put_status(:conflict)
        |> json(%{error: "Cannot transition from #{from} to #{to}"})

      {:error, :only_winning_contractor} ->
        conn |> put_status(:forbidden) |> json(%{error: "Only the winning contractor can start work"})

      {:error, :only_homeowner_can_cancel} ->
        conn |> put_status(:forbidden) |> json(%{error: "Only the homeowner can cancel"})

      {:error, reason} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: reason})
    end
  rescue
    ArgumentError ->
      conn |> put_status(:bad_request) |> json(%{error: "Invalid status"})
  end

  def accept_bid(conn, %{"id" => job_id, "bid_id" => bid_id}) do
    case Marketplace.accept_bid(job_id, bid_id) do
      {:ok, bid} ->
        json(conn, %{bid: serialize_bid(bid)})

      {:error, :bid_not_found} ->
        conn |> put_status(:not_found) |> json(%{error: "Bid not found"})

      {:error, reason} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: reason})
    end
  end

  defp serialize_job(job) do
    %{
      id: job.id,
      title: job.title,
      description: job.description,
      category: job.category,
      budget_min: job.budget_min,
      budget_max: job.budget_max,
      location: job.location,
      status: job.status,
      bid_count: job.bid_count,
      homeowner: serialize_user(job.homeowner),
      posted_at: job.inserted_at
    }
  end

  defp serialize_bid(bid) do
    %{
      id: bid.id,
      job_id: bid.job_id,
      amount: bid.amount,
      message: bid.message,
      timeline: bid.timeline,
      status: bid.status,
      contractor: serialize_user(bid.contractor),
      placed_at: bid.inserted_at
    }
  end

  defp serialize_user(%{id: _} = user) do
    %{id: user.id, name: user.name, role: user.role}
  end

  defp serialize_user(_), do: nil

  defp format_errors(%Ecto.Changeset{} = changeset) do
    Ecto.Changeset.traverse_errors(changeset, fn {msg, opts} ->
      Regex.replace(~r"%{(\w+)}", msg, fn _, key ->
        opts |> Keyword.get(String.to_existing_atom(key), key) |> to_string()
      end)
    end)
  end

  defp format_errors(reason), do: reason
end

File: src/main/kotlin/com/strata/ftw/web/controller/JobController.kt
package com.strata.ftw.web.controller

import com.strata.ftw.domain.entity.JobStatus
import com.strata.ftw.service.MarketplaceService
import com.strata.ftw.service.TokenClaims
import com.strata.ftw.web.dto.*
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import java.util.UUID

@RestController
@RequestMapping("/api/jobs")
class JobController(private val marketplace: MarketplaceService) {

    @GetMapping
    fun listJobs(
        @RequestParam status: String?,
        @RequestParam category: String?,
        @RequestParam limit: Int?
    ): ResponseEntity<Any> {
        val jobStatus = status?.let { JobStatus.valueOf(it) }
        val jobs = marketplace.listJobs(jobStatus, category, limit ?: 20)
        return ResponseEntity.ok(mapOf("jobs" to jobs.map { marketplace.serializeJob(it) }))
    }

    @GetMapping("/{id}")
    fun getJob(@PathVariable id: UUID): ResponseEntity<Any> {
        val job = marketplace.getJob(id) ?: return ResponseEntity.notFound().build()
        val bids = marketplace.listBids(id)
        return ResponseEntity.ok(mapOf(
            "job" to marketplace.serializeJob(job),
            "bids" to bids.map { marketplace.serializeBid(it) }
        ))
    }

    @PostMapping
    fun createJob(
        @Valid @RequestBody req: CreateJobRequest,
        @AuthenticationPrincipal claims: TokenClaims
    ): ResponseEntity<Any> {
        val attrs = mapOf<String, Any?>(
            "title" to req.title,
            "description" to req.description,
            "category" to req.category,
            "budget_min" to req.budget_min,
            "budget_max" to req.budget_max,
            "location" to req.location
        ).filterValues { it != null }.mapValues { it.value!! }
        val job = marketplace.postJob(attrs, claims.userId)
        return ResponseEntity.status(HttpStatus.CREATED).body(mapOf("job" to marketplace.serializeJob(job)))
    }

    @PostMapping("/{jobId}/bids")
    fun placeBid(
        @PathVariable jobId: UUID,
        @Valid @RequestBody req: PlaceBidRequest,
        @AuthenticationPrincipal claims: TokenClaims
    ): ResponseEntity<Any> {
        val attrs = mapOf<String, Any?>(
            "amount" to req.amount,
            "message" to req.message,
            "timeline" to req.timeline
        ).filterValues { it != null }.mapValues { it.value!! }
        val bid = marketplace.placeBid(jobId, attrs, claims.userId)
        return ResponseEntity.status(HttpStatus.CREATED).body(mapOf("bid" to marketplace.serializeBid(bid)))
    }

    @PostMapping("/{jobId}/bids/{bidId}/accept")
    fun acceptBid(
        @PathVariable jobId: UUID,
        @PathVariable bidId: UUID,
        @AuthenticationPrincipal claims: TokenClaims
    ): ResponseEntity<Any> {
        val bid = marketplace.acceptBid(jobId, bidId, claims.userId)
        return ResponseEntity.ok(mapOf("bid" to marketplace.serializeBid(bid)))
    }

    @PostMapping("/{jobId}/transition")
    fun transitionJob(
        @PathVariable jobId: UUID,
        @Valid @RequestBody req: TransitionJobRequest,
        @AuthenticationPrincipal claims: TokenClaims
    ): ResponseEntity<Any> {
        val newStatus = JobStatus.valueOf(req.status)
        val job = marketplace.transitionJob(jobId, newStatus, claims.userId)
        return ResponseEntity.ok(mapOf("job" to marketplace.serializeJob(job)))
    }
}
